ScriptEngine fails to init in 2.12.3 with ClassPath

```
$ ~/scala-2.12.2/bin/scala
Welcome to Scala 2.12.2 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_111).
Type in expressions for evaluation. Or try :help.

scala> import javax.script._
import javax.script._

scala> val saved0 = Thread.currentThread.getContextClassLoader
saved0: ClassLoader = scala.tools.nsc.interpreter.IMain$TranslatingClassLoader@3a71c100

scala> val scriptEngineManager = new ScriptEngineManager()
scriptEngineManager: javax.script.ScriptEngineManager = javax.script.ScriptEngineManager@681e144

scala> val se = scriptEngineManager.getEngineByName("scala")
se: javax.script.ScriptEngine = scala.tools.nsc.interpreter.Scripted@1ddc6db2
```
but
```
$ scala

     ________ ___   / /  ___  
    / __/ __// _ | / /  / _ | 
  __\ \/ /__/ __ |/ /__/ __ | 
 /____/\___/_/ |_/____/_/ | | 
                          |/  version 2.12.3

scala> import javax.script._
import javax.script._

scala> val saved0 = Thread.currentThread.getContextClassLoader
saved0: ClassLoader = scala.tools.nsc.interpreter.IMain$TranslatingClassLoader@6719a5b8

scala> val scriptEngineManager = new ScriptEngineManager()
scriptEngineManager: javax.script.ScriptEngineManager = javax.script.ScriptEngineManager@3421debd

scala> val se = scriptEngineManager.getEngineByName("scala")
java.lang.AssertionError: assertion failed: Directory file in DirectoryFileLookup cannot be null
  at scala.tools.nsc.classpath.JFileDirectoryLookup.$init$(DirectoryClassPath.scala:120)
  at scala.tools.nsc.classpath.DirectoryClassPath.<init>(DirectoryClassPath.scala:202)
  at scala.tools.nsc.classpath.ClassPathFactory$.newClassPath(ClassPathFactory.scala:78)
  at scala.tools.nsc.classpath.ClassPathFactory.newClassPath(ClassPathFactory.scala:20)
  at scala.tools.nsc.classpath.ClassPathFactory.$anonfun$classesInManifest$1(ClassPathFactory.scala:49)
  at scala.tools.nsc.classpath.ClassPathFactory.classesInManifest(ClassPathFactory.scala:49)
  at scala.tools.util.PathResolver$Calculated$.basis(PathResolver.scala:244)
  at scala.tools.util.PathResolver$Calculated$.containers$lzycompute(PathResolver.scala:248)
  at scala.tools.util.PathResolver$Calculated$.containers(PathResolver.scala:248)
  at scala.tools.util.PathResolver.containers(PathResolver.scala:264)
  at scala.tools.util.PathResolver.computeResult(PathResolver.scala:286)
  at scala.tools.util.PathResolver.result(PathResolver.scala:269)
  at scala.tools.nsc.backend.JavaPlatform.classPath(JavaPlatform.scala:23)
  at scala.tools.nsc.backend.JavaPlatform.classPath$(JavaPlatform.scala:22)
  at scala.tools.nsc.Global$GlobalPlatform.classPath(Global.scala:106)
  at scala.tools.nsc.Global.classPath(Global.scala:117)
  at scala.tools.nsc.Global$GlobalMirror.rootLoader(Global.scala:58)
  at scala.reflect.internal.Mirrors$Roots$RootClass.<init>(Mirrors.scala:276)
  at scala.reflect.internal.Mirrors$Roots.RootClass$lzycompute(Mirrors.scala:290)
  at scala.reflect.internal.Mirrors$Roots.RootClass(Mirrors.scala:290)
  at scala.reflect.internal.Mirrors$Roots$EmptyPackageClass.<init>(Mirrors.scala:299)
  at scala.reflect.internal.Mirrors$Roots.EmptyPackageClass$lzycompute(Mirrors.scala:305)
  at scala.reflect.internal.Mirrors$Roots.EmptyPackageClass(Mirrors.scala:305)
  at scala.reflect.internal.Mirrors$Roots.EmptyPackageClass(Mirrors.scala:245)
  at scala.reflect.internal.Mirrors$RootsBase.init(Mirrors.scala:219)
  at scala.tools.nsc.Global.rootMirror$lzycompute(Global.scala:65)
  at scala.tools.nsc.Global.rootMirror(Global.scala:63)
  at scala.tools.nsc.Global.rootMirror(Global.scala:36)
  at scala.reflect.internal.Definitions$DefinitionsClass.ObjectClass$lzycompute(Definitions.scala:267)
  at scala.reflect.internal.Definitions$DefinitionsClass.ObjectClass(Definitions.scala:267)
  at scala.reflect.internal.Definitions$DefinitionsClass.init(Definitions.scala:1448)
  at scala.tools.nsc.Global$Run.<init>(Global.scala:1154)
  at scala.tools.nsc.interpreter.IMain._initialize(IMain.scala:125)
  at scala.tools.nsc.interpreter.IMain.initializeSynchronous(IMain.scala:147)
  at scala.tools.nsc.interpreter.Scripted.<init>(Scripted.scala:74)
  at scala.tools.nsc.interpreter.Scripted$.apply(Scripted.scala:309)
  at scala.tools.nsc.interpreter.Scripted$Factory.getScriptEngine(Scripted.scala:302)
  at javax.script.ScriptEngineManager.getEngineByName(ScriptEngineManager.java:238)
  ... 29 elided

```

This is what i get no matter what library. I would like to start another ticket but since every search leads to this one perhaps it will be beneficial for future victims.

java.lang.AssertionError: assertion failed: Directory file in DirectoryFileLookup cannot be null
	at scala.tools.nsc.classpath.JFileDirectoryLookup.$init$(DirectoryClassPath.scala:120)
	at scala.tools.nsc.classpath.DirectoryClassPath.<init>(DirectoryClassPath.scala:280)
	at scala.tools.nsc.classpath.ClassPathFactory$.newClassPath(ClassPathFactory.scala:78)
	at scala.tools.nsc.classpath.ClassPathFactory.newClassPath(ClassPathFactory.scala:20)
	at scala.tools.nsc.classpath.ClassPathFactory.$anonfun$classesInManifest$1(ClassPathFactory.scala:49)
	at scala.tools.nsc.classpath.ClassPathFactory.classesInManifest(ClassPathFactory.scala:49)
	at scala.tools.util.PathResolver$Calculated$.basis(PathResolver.scala:253)
	at scala.tools.util.PathResolver$Calculated$.containers$lzycompute(PathResolver.scala:257)
	at scala.tools.util.PathResolver$Calculated$.containers(PathResolver.scala:257)
	at scala.tools.util.PathResolver.containers(PathResolver.scala:273)
	at scala.tools.util.PathResolver.computeResult(PathResolver.scala:295)
	at scala.tools.util.PathResolver.result(PathResolver.scala:278)
	at scala.tools.nsc.backend.JavaPlatform.classPath(JavaPlatform.scala:23)
	at scala.tools.nsc.backend.JavaPlatform.classPath$(JavaPlatform.scala:22)
	at scala.tools.nsc.Global$GlobalPlatform.classPath(Global.scala:113)
	at scala.tools.nsc.Global.classPath(Global.scala:124)
	at scala.tools.nsc.Global$GlobalMirror.rootLoader(Global.scala:59)
	at scala.reflect.internal.Mirrors$Roots$RootClass.<init>(Mirrors.scala:274)
	at scala.reflect.internal.Mirrors$Roots.RootClass$lzycompute(Mirrors.scala:288)
	at scala.reflect.internal.Mirrors$Roots.RootClass(Mirrors.scala:288)
	at scala.reflect.internal.Mirrors$Roots$EmptyPackageClass.<init>(Mirrors.scala:297)
	at scala.reflect.internal.Mirrors$Roots.EmptyPackageClass$lzycompute(Mirrors.scala:303)
	at scala.reflect.internal.Mirrors$Roots.EmptyPackageClass(Mirrors.scala:303)
	at scala.reflect.internal.Mirrors$Roots.EmptyPackageClass(Mirrors.scala:243)
	at scala.reflect.internal.Mirrors$RootsBase.init(Mirrors.scala:217)
	at scala.tools.nsc.Global.rootMirror$lzycompute(Global.scala:66)
	at scala.tools.nsc.Global.rootMirror(Global.scala:64)
	at scala.tools.nsc.Global.rootMirror(Global.scala:37)
	at scala.reflect.internal.Definitions$DefinitionsClass.ObjectClass$lzycompute(Definitions.scala:286)
	at scala.reflect.internal.Definitions$DefinitionsClass.ObjectClass(Definitions.scala:286)
	at scala.reflect.internal.Definitions$DefinitionsClass.init(Definitions.scala:1529)
	at scala.tools.nsc.Global$Run.<init>(Global.scala:1179)
	at scala.tools.nsc.interpreter.IMain.liftedTree1$1(IMain.scala:137)
	at scala.tools.nsc.interpreter.IMain.global$lzycompute(IMain.scala:136)
	at scala.tools.nsc.interpreter.IMain.global(IMain.scala:127)
	at scala.tools.nsc.interpreter.IMain.initializeCompiler(IMain.scala:125)
	at scala.tools.nsc.interpreter.shell.Scripted.<init>(Scripted.scala:30)
	at scala.tools.nsc.interpreter.shell.Scripted$.apply(Scripted.scala:268)
	at scala.tools.nsc.interpreter.shell.Scripted$Factory.getScriptEngine(Scripted.scala:261)
	at javax.script.ScriptEngineManager.getEngineByName(Unknown Source)
	at Blockchain.core.logic.startapps(logic.java:843)
	at Blockchain.core.logic.entrypoint(logic.java:1414)
	at Blockchain.core.logic.init(logic.java:1236)
	at Blockchain.start.main(start.java:160)

--- a/src/compiler/scala/tools/nsc/classpath/FileUtils.scala
+++ b/src/compiler/scala/tools/nsc/classpath/FileUtils.scala
@@ -15,7 +15,7 @@ package scala.tools.nsc.classpath
 import java.io.{File => JFile, FileFilter}
 import java.net.URL
 import scala.reflect.internal.FatalError
-import scala.reflect.io.AbstractFile
+import scala.reflect.io.{AbstractFile, ZipArchive}
 
 /**
  * Common methods related to Java files and abstract files used in the context of classpath
@@ -29,7 +29,7 @@ object FileUtils {
     def isScalaOrJavaSource: Boolean = !file.isDirectory && (file.hasExtension("scala") || file.hasExtension("java"))
 
     // TODO do we need to check also other files using ZipMagicNumber like in scala.tools.nsc.io.Jar.isJarOrZip?
-    def isJarOrZip: Boolean = file.hasExtension("jar") || file.hasExtension("zip")
+    def isJarOrZip: Boolean = file.hasExtension("jar") || file.hasExtension("zip") || file.isInstanceOf[ZipArchive]
 
     /**
      * Safe method returning a sequence containing one URL representing this file, when underlying file exists,

